keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,575 @@
+mechanism since some defaulting still requires non-static values that
+need to be computed and can't be defined declaratively. We could replace
+some defaulting that doesn't require to be done as code though.
+
+## Proposal
+
+It is important to us that the OpenAPI definition of a type should
+result in the same behavior, whether the type is built-in or a CRD.
+
+In order to do that, we need to change how CRDs are defaulted, and
+create a defaulting mechanism for built-in that would behave the same
+way.
+
+### CRD Normalization
+
+Null values in CRDs are currently not defaulted. On the other hand,
+`null` typically means "Devoid of
+value"(link)[https://yaml.org/type/null.html] in json/yaml, though it is
+treated as different from not present ("a null value is valid and
+different from not having that key in the mapping"). JSon Merge patch
+interprets a "null" value for a field as a means to delete it: "Null
+values in the merge patch are given special meaning to indicate the
+removal of existing values in the
+target."(link)[https://tools.ietf.org/rfc/rfc7386.txt#:~:text=Null%20values%20in%20the%20merge]
+
+On the other hand, while `{field: null}` and `{}` probably have the same
+meaning, they do not "look" the same and are hard to compare for
+clients. They are also currently treated differently by Kubernetes for
+defaulting purposes.
+
+We propose the creation of a new CRD flag: `preserveNullValues` which
+will default to `true`. This flag will be used as a ratcheting mechanism
+as we migrate to this new semantic throughout all Kubernetes types
+(which would be how all types eventually work). When false, the
+apiserver will use new semantics and remove all null values from CR
+objects. When true, the existing and deprecated semantics will be kept.
+That new flag is following the same logic as "preserveUnknownFields" and
+the new operation will happen using the same mechanisms right after
+pruning.
+
+Note that we still consider `{}` (and 0, 0.0, "", false, []) to be
+different from `null`, so `{foo: {}}` is different from `{}` or from
+`{foo: null}` (but the latter two are considered equal).
+
+As described by @sttts on many occasions, the immutability is blocked
+because we don't know how to distinguish between null and
+unspecified. We believe that this mechanism will be able to unblock it.
+
+### Marker format
+
+The marker will have the following format:
+```
+// +default=<any>
+```
+which describes the value that the server will use if the field has a
+zero-value.
+
+`<any>` has to be in one-line YAML format (including its json form),
+representing the actual value to use by default. That is different from
+the current encoding in kubebuilder, but there seems to be consensus
+that they would be able to change that to YAML too for that new marker.
+
+Namely, this will allow to preserve the concise semantics that you want
+for most values:
+
+```golang
+type Foo struct {
+  // +default=32
+  Integer int
+
+  // +default=bar
```

Comment:
Can we spec WRT whitespace and escapes?  Can I have a value with a newline in it? Tab?  Space?

Reply:
https://play.golang.org/p/kxNxW48GJ7Z
```golang
package main

import (
	"encoding/json"
	"fmt"
)

func main() {
	var i interface{}
	err := json.Unmarshal([]byte(`{" ": "tab:\t\nspace: "}`), &i)
	if err != nil {
		panic(err)
	}
	fmt.Println(i)
}
```